src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java

```diff
@@ -3816,16 +3816,28 @@ public JCTree.JCCompilationUnit parseCompilationUnit() {
             defs.append(pd);
         }
 
-        boolean checkForImports = true;
-        boolean firstTypeDecl = true;
+        boolean firstTypeDecl = true;   // have we see a class, enum, or interface declaration yet?
         while (token.kind != EOF) {
             if (token.pos <= endPosTable.errorEndPos) {
                 // error recovery
-                skip(checkForImports, false, false, false);
+                skip(firstTypeDecl, false, false, false);
                 if (token.kind == EOF)
                     break;
             }
-            if (checkForImports && mods == null && token.kind == IMPORT) {
+            // JLS 7.3 doesn't allow extra semi-colons after package or import declarations,
+            // but here we try to provide a more helpful error message if we encounter any.
+            // Do that by slurping in as many semi-colons as possible, and then seeing what
+            // comes after before deciding how best to handle them.
```

Comment:
Maybe it's my fault for misleading you in the discussion, but the dictionary says that "semicolon" is not a hyphenated word.

Reply:
> Maybe it's my fault for misleading you in the discussion, but the dictionary says that "semicolon" is not a hyphenated word.

Nope, totally my fault. Apparently I've been spelling it wrong for years. Will fix.